### PR TITLE
"Loop labels" greedy decoding: faster implementation

### PR DESCRIPTION
**Pull request by artbataev**
# What does this PR do ?

Improves "loop labels" greedy decoding implementation, introduced in #7926, with using non-blocking operations (avoiding device-to-host synchronization).

On my local machine, the time for decoding unsorted `test-other` (LibriSpeech) with `stt_en_fastconformer_transducer_large` (2 repeated measurements, in seconds; after warmup the results are better):

 bs | precision  | RNN-T, loop_labels v1 (main) | RNN-T, loop_labels v2 (this)
-- | -- | -- | --
16 | fp32 | 50s / 50s | 37s / 36s
128 | fp32 | 34s / 33s | 30s / 29s
16 | bf16 | 48s / 46s | 34s / 32s
128 | bf16 | 25s / 24s | 21s / 20s

This algorithm is planned for further improvements with torch.jit.script in #8203

**Collection**: [ASR]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage

```shell
python examples/asr/speech_to_text_eval.py \
   model_path=<nemo_model.nemo> \
   dataset_manifest=<manifest> \
   batch_size=128 \
   output_filename=<output_mainfest_path> \
   rnnt_decoding.greedy.loop_labels=true
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

**Comment by artbataev** [comment]
jenkins

File: nemo/collections/asr/modules/rnnt.py [hunk 1]
```diff
@@ -368,6 +368,20 @@ def batch_concat_states(self, batch_states: List[List[torch.Tensor]]) -> List[to
 
         return state_list
 
+    def batch_replace_states_mask(
+        self, src_states: list[torch.Tensor], dst_states: list[torch.Tensor], mask: torch.Tensor,
+    ):
+        """Replace states in dst_states with states from src_states using the mask"""
+        # same as `dst_states[0][mask] = src_states[0][mask]`, but non-blocking
+        torch.where(mask.unsqueeze(-1), src_states[0], dst_states[0], out=dst_states[0])
```
Comment (GNroy) [inline nemo/collections/asr/modules/rnnt.py:376]: Please state somehow that `batch_replace_states_mask` is an in-place method.

File: nemo/collections/asr/modules/rnnt.py [hunk 2]
```diff
@@ -1030,6 +1045,28 @@ def batch_concat_states(self, batch_states: List[List[torch.Tensor]]) -> List[to
 
         return state_list
 
+    def batch_replace_states_mask(
```
Comment (GNroy) [inline nemo/collections/asr/modules/rnnt.py:1048]: Same here

File: nemo/collections/asr/modules/rnnt_abstract.py [hunk 3]
```diff
@@ -278,6 +278,19 @@ def batch_select_state(self, batch_states: List[torch.Tensor], idx: int) -> List
         """
         raise NotImplementedError()
 
+    def batch_replace_states_mask(
```
Comment (GNroy) [inline nemo/collections/asr/modules/rnnt_abstract.py:281]: Same

**Review by GNroy — DISMISSED** [review]
LGTM, see comments.

File: nemo/collections/asr/modules/rnnt.py (hunk 1 above)
Comment (artbataev) [reply nemo/collections/asr/modules/rnnt.py:376]: Hm, can you please explain how to do this? This is described in the docstring.

Comment (GNroy) [reply nemo/collections/asr/modules/rnnt.py:376]: `def batch_replace_states_mask_(...)`
```
"""Replace states in dst_states with states from src_states using the mask.
In-place operation."""
```

Comment (artbataev) [reply nemo/collections/asr/modules/rnnt.py:376]: I think this style is not appropriate here: this is used for in-place modification when calling object methods, e.g. 

```python
x = torch.rand(...)
y = torch.rand(...)
x.logical_and_(y) # in-place
```

Since `batch_replace_states_mask` does not modify the **`self`** state, it should not be marked with underscore (it modifies the object passed, not self).

Cmp: `x.logical_and_(y)` (modify `x`) and `torch.logical_and(x, y, out=x)` (`torch` is not modified, but `x` is still modified)

Comment (GNroy) [reply nemo/collections/asr/modules/rnnt.py:376]: About `self`: why do we need `self` in the definition if nothing is modified/used?

Comment (artbataev) [reply nemo/collections/asr/modules/rnnt.py:376]: Hmm, maybe we speak about different things. 
I mean, that `batch_replace_states_mask` is not an in-place method (and should not be marked with underscore), since it does not modify the caller (self).

Comment (GNroy) [reply nemo/collections/asr/modules/rnnt.py:376]: > I mean, that batch_replace_states_mask is not an in-place method (and should not be marked with underscore), since it does not modify the caller (self).

I agree. My question is why do we need self here since it is not used whatsoever?

Comment (artbataev) [reply nemo/collections/asr/modules/rnnt.py:376]: Ah, do you mean to make the method static/classmethod?

Comment (GNroy) [reply nemo/collections/asr/modules/rnnt.py:376]: Yep

Comment (artbataev) [reply nemo/collections/asr/modules/rnnt.py:376]: Ah, ok, thanks! I will fix it)

**Comment by artbataev** [comment]
jenkins

File: nemo/collections/asr/modules/rnnt.py (hunk 1 above)
Comment (artbataev) [reply nemo/collections/asr/modules/rnnt.py:377]: Done, made a classmethod.
Sorry for the misunderstanding!

**Review by titu1994 — APPROVED** [review]
Really incredible work ! Ready to merge

**Comment by titu1994** [comment]
jenkins
